packages/server/src/artifact/checkin/output.rs

```diff
@@ -631,11 +631,24 @@ impl Server {
 			return Ok(());
 		};
 		visited.insert(path.clone());
+
 		let (data, id) = {
 			let output = output.read().unwrap();
 			(output.data.clone(), output.id.clone())
 		};
 
+		// macOS disables hardlinking for some files within .app dirs, such as *.app/Contents/{PkgInfo,Resources/\*.lproj,_CodeSignature} and .DS_Store.
+		// Perform a copy instead in these cases.
+		// See <https://github.com/NixOS/nix/blob/95f2b2beabc1ab0447bb4683516598265d71b695/src/libstore/optimise-store.cc#L100>
+		let should_hardlink = if cfg!(target_os = "macos") {
+			let app_dir_re: std::sync::LazyLock<regex::Regex> =
+				std::sync::LazyLock::new(|| regex::Regex::new(r"\.app/.*Contents/.+$").unwrap());
```

Comment:
Just a note - the correct regex here is `r"\.app/Contents/.+$"`, i.e. the `Contents` directory is a direct child of the `*.app` directory. The tip of `main` currently has a bug which is inserting current components in paths, leading to paths like `foo.app/./Contents/./bar`, so this iteration allows characters in between the .app components and the Contents component. Once the path changes are landed, I will update the regex and merge.